arwen/contexts/metering.go

```diff
@@ -50,8 +51,8 @@ func (context *meteringContext) SetGasSchedule(gasMap config.GasScheduleMap) {
 
 // UseGas sets in the runtime context the given gas as gas used
 func (context *meteringContext) UseGas(gas uint64) {
-	gasUsed := context.host.Runtime().GetPointsUsed() + gas
-	context.host.Runtime().SetPointsUsed(gasUsed)
+	gasUsed := math.AddInt64(int64(context.host.Runtime().GetPointsUsed()), int64(gas))
```

Comment:
Create a variable `gas_i64` and set it to `MaxInt64` (signed) in case the `gas` argument (unsigned) would overflow during the conversion `int64(gas)`. Otherwise, a contract that somehow managed to force a situation with `metering.UseGas(MaxInt64 + 10)` would actually regain `9` gas due to the u64 → i64 overflow. 

This is a hypothetical issue, and is not expected to occur in practice.

Reply:
Same would go for `context.host.Runtime().GetPointsUsed()` - it's the same u64 → i64 conversion. 

However, these conversions should probably be handled in the Runtime context, in the methods `GetPointsUsed()` and `SetPointsUsed()`. Can you please add conditions in these two methods, to ensure that they stay within the bounds of i64? It's safest.

Reply:
Added protection for SetPointsUsed(), and used Uint64 operations everywhere else.